js/reference.js

```diff
@@ -4350,75 +4428,266 @@
-
+    function FacetColumn (reference, index, column, json, filters) {
+        
         /**
-         * The {@link ERMrest.ReferenceColumn} that this facet belongs to. If it is
-         * undefined, that means this facet belongs to the table and not column.
-         * @type {ERMrest.ReferenceColumn}
+         * The column object that the filters are based on
+         * @type {ERMrest.Column}
+         */
+        this._column = column;
+        
+        /**
+         * [reference description]
+         * @type {ERMrest.Reference}
          */
-        this.column = refColumn;
-
         this.reference = reference;
-
+        
+        /**
+         * The index of facetColumn in the list of facetColumns
+         * NOTE: Might not be needed
+         * @type {int}
+         */
         this.index = index;
 
         /**
+         * A valid data-source path
+         * NOTE: we're not validating this data-source, we assume that this is valid.
+         * @type {obj|string} 
+         */
+        this.dataSource = json.source;
+        
+        /**
          * Filters that are applied to this facet.
+         * @type{FacetFilter[]}
          */
         this.filters = [];
+        if (Array.isArray(filters)) {
+            this.filters = filters;
+        } else {
+            this.setFilters(json);
+        }
+        
+        // the whole filter object
+        this._json = json;
     }
     FacetColumn.prototype = {
         constructor: FacetColumn,
-
+        
+        // returns the last foreignkey object in the path
+        get _lastForeignKey() {
+            if (this._lastForeignKey_cached === undefined) {
+                if (!Array.isArray(this.dataSource)) {
+                    this._lastForeignKey_cached = null;
+                } else {
+                    var lastJoin = this.dataSource[this.dataSource.length-2];
+                    var isInbound = false, constraint;
+                    
+                    if ("inbound" in lastJoin) {
+                        isInbound = true;
+                        constraint = lastJoin.inbound;
+                    } else {
+                        constraint = lastJoin.outbound;
+                    }
+                    
+                    this._lastForeignKey_cached = {
+                        "obj": module._getConstraintObject(this._column.table.schema.catalog.id, constraint[0], constraint[1]).object,
+                        "isInbound": isInbound
+                    };
+                }
+            }
+            return this._lastForeignKey_cached;
+        },
+        
         /**
-         * data source of this facet. Can be used to find filters for this facet
-         * @return {String} [description]
+         * The Preferred ux mode.
+         * Any of:
+         * `choices`, `range`, or `search`
+         * This should be used if we're not in entity mode.
+         * TODO: what should be the default? This will eventually change,
```

Comment:
If we want to consider a default mode for facets for any column type, we might want to keep it simple and have the default mode show as `choices`. `Search` mode would imply that the user needs to be aware of the whole set of values they are searching through. Choices provides some of that information for them.